fix similar() and axes() for BorderArray

Cody-G opened the pull request:
Addresses #137

Cody-G commented:
I also rolled in a fix for #139

johnnychen94 commented on src/borderarray.jl:
```diff
@@ -91,7 +91,7 @@ end
 @inline function _outeraxes(arr, border)
     map(axes(arr), border.lo, border.hi) do r, lo, hi
-        (first(r)-lo):(last(r)+hi)
+        Base.IdentityUnitRange((first(r)-lo):(last(r)+hi))
```
I'm not familiar with the code base thus I don't know if this change is expected.

But to keep compatibility to julia v1.0, you could add a custom implementation
```julia
if VERSION < v"1.1"
    IdentityUnitRange = Base.IdentityUnitRange
else
    IdentityUnitRange = # self-defined
end
```

timholy replied:
The best portability option is `OffsetArrays.IdentityUnitRange` which is defined here: https://github.com/JuliaArrays/OffsetArrays.jl/blob/e22f56ed7f80048ebe9662eeae7dc2891ecf0b8d/src/OffsetArrays.jl#L6-L10

jw3126 commented on test/border.jl:
```diff
@@ -270,6 +270,13 @@ using Test
+        b_front = BorderArray(zeros(10), Fill(1.0, (3,), (0,)))
+        b_both = BorderArray(b_front, Fill(1.0, (0,), (5,)))
+        b_sim = similar(b_both)
+        @test axes(b_sim) == axes(b_both)
```
Can you maybe also add a few very explicit `axes` tests? Like
`@test axes(b) === (IdentityUnitRange(...)),)`?

timholy replied:
Or rather than testing the type, maybe best would be to test with `==`, which should test both the values of the axes and the axes-of-the-axes.

Cody-G replied:
I think I'm already testing with `==` rather than testing the type?  So I'm not sure what you mean here.

timholy replied:
Don't just test that they are the same, test the actual value produced. E.g., if you expect `f(x)` to return 1, testing `f(x) == f(y)` will not catch certain classes of bugs.

Cody-G replied:
Ah got it, thanks!

Cody-G replied:
> Can you maybe also add a few very explicit `axes` tests? Like
> `@test axes(b) === (IdentityUnitRange(...)),)`?

I added that test, just let me know if there are others you'd like added

jw3126 commented on src/borderarray.jl:
```diff
@@ -74,7 +74,7 @@ function Base.similar(A::BorderArray, ::Type{T}, dims::Dims) where T
 end
 const OffsetAxis = Union{Integer, UnitRange, Base.OneTo, IdentityUnitRange}
 function Base.similar(A::BorderArray, ::Type{T}, inds::Tuple{OffsetAxis,Vararg{OffsetAxis}}) where T
```
Can you add tests for this method specifically?

Cody-G commented:
Thanks for the reviews!  I've addressed the comments and rebased.  Just let me know if there are additional ways that I should be testing `axes()` and `similar()`, I'm not sure if I covered everything.

timholy commented on src/borderarray.jl:
```diff
@@ -91,7 +91,7 @@ end
 @inline function _outeraxes(arr, border)
     map(axes(arr), border.lo, border.hi) do r, lo, hi
-        (first(r)-lo):(last(r)+hi)
+        Base.IdentityUnitRange((first(r)-lo):(last(r)+hi))
```
```suggestion
        OffsetArrays.IdentityUnitRange((first(r)-lo):(last(r)+hi))
```

Cody-G replied:
Oh shoot, I thought I switched that, sorry.  Will fix now

Cody-G replied:
Should be fixed with the latest push (`OffsetArrays.IdentityUnitRange` was already imported so I didn't need any explicit prefix)

timholy commented:
Thanks, @Cody-G!
